wgpu-native/src/device.rs

```diff
@@ -113,20 +120,48 @@ impl<B: hal::Backend> DestroyedResources<B> {
                 assert_eq!(num_refs, 2);
                 let resource_id = self.referenced.swap_remove(i).0;
                 let (submit_index, resource) = match resource_id {
-                    ResourceId::Buffer(id) => {
+                    ResourceOperationById::DeleteBuffer(id) => {
                         #[cfg(feature = "local")]
                         HUB.buffers.unregister(id);
                         let buf = buffer_guard.take(id);
                         let si = buf.life_guard.submission_index.load(Ordering::Acquire);
-                        (si, Resource::Buffer(buf))
+                        (si, ResourceOperation::DeleteBuffer(buf))
                     }
-                    ResourceId::Texture(id) => {
+                    ResourceOperationById::DeleteTexture(id) => {
                         #[cfg(feature = "local")]
                         HUB.textures.unregister(id);
                         let tex = texture_guard.take(id);
                         let si = tex.life_guard.submission_index.load(Ordering::Acquire);
-                        (si, Resource::Texture(tex))
+                        (si, ResourceOperation::DeleteTexture(tex))
+                    }
+                    _ => unimplemented!(),
+                };
+                match self
+                    .active
+                    .iter_mut()
+                    .find(|a| a.index == submit_index)
+                {
+                    Some(a) => a.resources.push(resource),
+                    None => self.free.push(resource),
+                }
+            }
+        }
+        for i in (0..self.mapped.len()).rev() {
+            // one in resource itself, one here in this list, one the owner holds, and one more somewhere?
+            let num_refs = self.mapped[i].1.load();
+            trace!("{} references remain", num_refs);
+            if num_refs <= 4 {
+                assert_eq!(num_refs, 4);
+                let resource_id = self.mapped.swap_remove(i).0;
+                let (submit_index, resource) = match resource_id {
+                    ResourceOperationById::MapBuffer(id) => {
+                        #[cfg(feature = "local")]
+                        HUB.buffers.unregister(id);
+                        let buf = buffer_guard.take(id);
```

Comment:
Obviously this is removing the buffer when it gets mapped, and hence breaking any future operations thereon. Ideally we could break out submission tracking from `DestroyedResources`, so that a separate `MappedResources` could also observe the same fence?